### PR TITLE
Tool Builder - Self Check Responses (FRONT-END)

### PR DESCRIPTION
**Pull request by jaybhagat**
## Notion ticket link

[[Tool Builder] Self Check Response](https://www.notion.so/uwblueprintexecs/Task-Board-485f846ae00243599dddba2be9bcba82?p=eefac0f1739049ce92abbf45673b9f6b)

## Implementation description
* Used array of breakpoints
* Each breakpoint has a component attached to it
* The component is created using the data from the breakpoint data in the array
* Creates the new breakpoint by setting the lower as the last breakpoint's upper + 1

## Steps to test
1. Open http://localhost:3000/admin/dashboard/tools
2. Create or edit an existing tool
3. Go to Self Check tab
4. Click Response
5. Explore! You can try to add new breakpoints, change the screen size, set the min and max of the breakpoints, add questions, etc. Note that you will have to create questions for the self check that are either multiple choice or multiselect to be able to play around with this functionality.

NOTE: the back-end will be done in another PR.

## What should reviewers focus on?
* Error checking
* Functions to compute the selected questions titles... Should I use an array with the useState hook to keep track of the number for the questions that are added? ~~Would have to sort these though~~ Just realized that I still have to sort them (or add them in sorted order)..
* Any other suggestions to make the code better! :)
* Let me know if there's something I forgot with this ticket!

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Review by ChinemeremChigbo — COMMENTED** [review]
image
Currently unable to set upper and lower for multiple choice like we talked about.

**Review by jennifer-lu — CHANGES_REQUESTED** [review]
Great work Jay! The UI is looking good. I just have a few concerns after testing.

For the first breakpoint, the minimum value displayed is NaN, I think that displaying 0 would be friendlier to users:
Screen Shot 2022-03-10 at 7 36 00 PM

There is a bug in this edge case:
1. Make a Self Check with two multiple choice questions
2. For the Self Check Score select All Self Check Questions
3. Go back to the question tab and make one of the questions short answer
4. In the Self Check Score section of the Response tab, All Self Check Questions should be selected but it isn't:
Screen Shot 2022-03-10 at 7 45 58 PM

I can't seem to edit the maximum value of a break point after I set it.

File: src/components/toolSelfCheckResponse/index.tsx [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                    to
+                </Text>
+                <Input
+                    bg="whiteAlpha"
+                    placeholder="Breakpoint Score"
+                    value={upperBound}
+                    textAlign="center"
+                    width="35%"
+                    display="inline-block"
+                    onBlur={(str) =>
```
Comment (jennifer-lu) [inline src/components/toolSelfCheckResponse/index.tsx:69]: Try changing `onBlur` to `onChange` to be able to edit the breakpoint score after unfocusing.

**Review by chamod-gamage — CHANGES_REQUESTED** [review]
Looking great!
Just one change to the functionality:
The options available for scoring should correspond to those questions which have a numeric data type. This includes all sliders, and any of the other question types for which the numeric data type has been selected. (e.g. 
image
)

**Comment by jaybhagat** [comment]
Hey @ChinemeremChigbo @jennifer-lu @chamod-gamage, I fixed up the bugs and added the suggestions you guys had! Feel free to take a look again :) (Hopefully it'll be all good this time).

**Review by chamod-gamage — CHANGES_REQUESTED** [review]
https://www.loom.com/share/188831d957144467b3f5c6c54a2696b6
Great work! Just a couple more quick fixes that I noted in this video^

Helpful SO link:
https://stackoverflow.com/questions/63224459/disable-scrolling-on-input-type-number-in-react

**Review by jennifer-lu — CHANGES_REQUESTED** [review]
The fixes are all looking good, nice work!

I'm seeing a bug in a similar edge case as before. It's pretty nitpicky, but would be nice to fix:
1. Make a Self Check with two multiple choice questions (numeric)
2. For the Self Check Score select All Self Check Questions
3. Go back to the question tab and delete the first question
4. In the Self Check Score section of the Response tab, one question is selected twice:

Screen Shot 2022-03-20 at 3 27 15 PM

5. Go back to the question tab an re-add the second numeric multiple choice
6. In the Self Check Score section of the Response tab, all questions are not selected but it shows they are:

Screen Shot 2022-03-20 at 3 41 44 PM

Another bug comes up in this case:
1. Make a Self Check with one multiple choice question (numeric)
2. Only fill out values of the first two options, leaving the third empty
3. For the Self Check Score select All Self Check Questions
4. Minimum and maximum scores show NaN:

Screen Shot 2022-03-20 at 3 51 03 PM

I have some product-related questions that maybe @chamod-gamage can answer:
1. Should users be able to delete breakpoints?
2. The minimum score for multiple choices seems to be the lowest option value. I think that using 0 instead might be better since a user might not answer the question.

**Review by jennifer-lu — COMMENTED** [review]
Please ignore this comment, just wanted to remove the request changes status. Oop unfortunately didn't work.

**Comment by jaybhagat** [comment]
> I have a product-related question that maybe @chamod-gamage can also answer. The minimum score for multiple choices seems to be the lowest option value. Would using 0 instead be better since a user might not answer the question?

@jennifer-lu talked with @chamod-gamage and Jenna about this. Seems like all the questions are required so I just left it like it is.

Also added a delete button not part of the designs but discussed with PMs.

**Review by jennifer-lu — COMMENTED** [review]
Looks amazing Jay! Thanks for adding the deletion feature and fixing the nitpicks. Everything is generally working well.

There seems to be a bug for the breakpoint keys:
1. Open the browser console
2. Set three breakpoints
3. Delete the second one
4. Add a third one and you'll see this in the console:
Screen Shot 2022-03-27 at 8 02 53 PM
5. Go to the questions tab and observe some unexpected behaviour (the delete breakpoint has reappeared):
Screen Shot 2022-03-27 at 8 04 02 PM

**Comment by jaybhagat** [comment]
@jennifer-lu @chamod-gamage just thought I should remind you guys about this PR! please review it whenever you get the time :)!

**Review by jennifer-lu — APPROVED** [review]
@jaybhagat Thanks for the reminder to review! This looks fantastic 🔥

I noticed a small nitpick:
1. Add a numeric multiple choice question with a max score of 5 (i.e. add option with corresponding value 5)
2. On the responses page select the question
3. On the question page make the question non-numeric
4. On the responses page the question is still selected and the maximum score is 5 despite the question being non-numeric:
Screen Shot 2022-04-06 at 12 49 41 PM
Screen Shot 2022-04-06 at 12 52 28 PM

File: pages/admin/dashboard/toolBuilder.tsx [hunk 2]
```diff
@@ -16,16 +23,22 @@ import {
...
+    SelfCheckResponseCard,
+    ToolHomePage,
+} from "@components";
 import { useRouter } from "next/router";
 import axios from "axios";
-import { useSWRConfig } from "swr";
 
 export const getServerSideProps: GetServerSideProps = async ({ req, res }) => {
-    if (process.env.USE_ADMIN_LOGIN === "true") {
+    if (process.env.NODE_ENV == "production") {
```
Comment (tawnyzhao) [inline pages/admin/dashboard/toolBuilder.tsx:41]: This change should be reverted

File: pages/admin/dashboard/toolBuilder.tsx [hunk 3]
```diff
@@ -170,6 +207,203 @@ const ToolBuilder: Page = () => {
         getAllTools();
     }, []);
 
+    const getMinSum = (questions) =>
```
Comment (tawnyzhao) [inline pages/admin/dashboard/toolBuilder.tsx:210]: Try to move these functions outside of the component itself you don't need to use any closure values

Comment (tawnyzhao) [reply pages/admin/dashboard/toolBuilder.tsx:210]: Any functions that need to interact with the component's state can stay inside the component body

File: pages/admin/dashboard/toolBuilder.tsx [hunk 4]
```diff
@@ -98,6 +121,20 @@ const ToolBuilder: Page = () => {
     const [lastSavedQuestions, setLastSavedQuestions] = useState(
         JSON.parse(JSON.stringify(defaultQuestions)),
     );
+    const [selectedQuestions, setSelectedQuestions] = useState([]);
+    const [description, setDescription] = useState("");
+    const [breakpoints, setBreakpoints] = useState([
```
Comment (tawnyzhao) [inline pages/admin/dashboard/toolBuilder.tsx:126]: consider defining a type for breakpoint in this file and pass it to useState<type>

**Review by tawnyzhao — APPROVED** [review]
Looks really good so far! a few requests for changes

**Review by ChinemeremChigbo — APPROVED** [review]
Nicely done!

**Review by chamod-gamage — APPROVED** [review]

**Pull request merged**